mdn/yari
# feat(lint): Implement front matter validator and linter

## Summary

- Fixes https://github.com/mdn/yari/issues/7900

### Problem

Front matter now hold many important and complex values. Without the automated validator it starts accumulating errors. And there is no tool to find and report/fix these issues.

Implement a custom linter that will satisfy following requirements:
- validation:
	- Make some attributes, like `title` and `slug`, mandatory.
	- Validate `page-type` attribute. And allow only certain values based on technology/feature of the document.
	- Enforce data types/formats/lengths. For example, `spec-urls` need to be URI format.
- prettier:
	- Indent yaml, remove trailing spaces, fix line widths etc.
	- Make sure attributes are in specific order: title -> slug -> page-type -> status -> browser-compat -> spec-urls
	- For attributes `browser-compat` and `spec-urls` support union data type `string|array`. If there is only one value then use string other wise use array for multivalue:
    ```yml
    # single value
    spec-urls: https://abcd.com/api.html

    # multiple values
    spec-urls:
      - https://abcd.com/api.html
      - https://xyz.com/property.html
    ```

- automation:
	- Make a PR check workflow that will flag validation errors. I.e., option to fix specified list of files.
	- Daily chron job workflow that will pretty and fix order. Just like the current daily markdownlint-fix workflow.

### Solution

Implement a custom solution similar to the filecheck tool.\
Convert front matter YAML to JSON object and validate using [JSON Schema](https://json-schema.org/draft/2020-12/json-schema-core.html) which also supports the required union type. Following are the required steps:
- define JSON schema for front matter. See `front-matter-config.json`.
- use `js-yaml` library to convert yaml to json
- use `ajv` library to validate the schema
- using custom logic validate the attribute order
- pretty print yaml using `js-yaml` library

To run the tool use following commands:
```bash
# check few files
yarn tool fmlint  "/.../index.md " "/.../index.md"

# pretty print and auto fix front matter
yarn tool fmlint --fix true "/.../index.md"

# check all files
yarn tool fmlint > ~/logs/fm_issues.txt

# auto pretty all files
yarn tool fmlint --fix true > ~/logs/fm_issues.txt
```

Let me know about following possiblities:
- move `front-matter-config.json` to `mdn/content` repo, so that configuration won't be attached to yari
- for each value a regular expression can be specified for tighter control
- maxLength for `title` is set to 150 chars in the PR, what is the required value?

## Screenshots

```bash
yarn tool fmlint /home/koder/GitHub/mdn/content/files/en-us/glossary/aria/index.md /home/koder/GitHub/mdn/content/files/en-us/web/svg/element/a/index.md > ~/_temp/fmfaults.txt 
progress [========================================] 100% | ETA: 0s | 2/2

error: Error: ../content/files/en-us/glossary/aria/index.md
       'page-type' property must be equal to one of the allowed values: 
       	glossary-definition, glossary-disambiguation
       
       Error: ../content/files/en-us/web/svg/element/a/index.md
       Front matter must have required property 'slug'
       'page-type' property must be equal to one of the allowed values: 
       	guide, landing-page
       
       Following fixable errors can be fixed using '--fix true' option
       ../content/files/en-us/web/svg/element/a/index.md
       	 Front matter attributes are not in required order: title->slug->page-type->status->browser-compat->spec-urls

error Command failed with exit code 1.                                     
```

## How did you test this change?

- The linter found these issues https://github.com/mdn/content/pull/24795/files
- It automatically removed unknown attribute `tag` using setting `additionalProperties: false` in the config. https://github.com/mdn/content/pull/24776/files
